transformers/generate/sampler.py

```diff
@@ -0,0 +1,210 @@
+            Device on which the computations will be run.
+    """
+
+    def __init__(self, config, device):
+        self.k = config.k
+        self.p = config.p
+        self.temperature = config.temperature
+        self.repetition_penalty = config.repetition_penalty
+
+        self.do_apply_temperature = True if config.temperature > 0 else False
+        self.do_apply_repetition_penalty = True if config.repetition_penalty > 1 else False
+
+        if self.p > 1:
+            warnings.warn(
+                "p is a probability; its value must lie between 0 and 1, got {}. Ignored.".format(
+                    self.p
+                )
+            )
+
+        self.device = device
+
+    def generate_sequence(self):
+        """ Generate a sequence of `length` tokens starting from the
+        provided `prompt`. This method is model-specific.
+        """
+        raise NotImplementedError
+
+    def generate_one_token(self, next_token_logits, past_sequence):
+        logits = self.apply_repetition_penalty(next_token_logits, past_sequence)
+        logits = self.apply_temperature(logits)
+        logits = self.apply_top_k_filter(logits)
+        logits = self.apply_nucleus_filter(logits)
+        return self.sample_one_token(logits)
+
+    def apply_repetition_penalty(self, logits, past_sequence):
+        """ Apply a penalty to tokens that appear more than once in the
+        generated sequence.
+
+        .. Keskar, Nitish Shirish, et al. "Ctrl: A conditional transformer
+           language model for controllable generation." arXiv preprint
+           arXiv:1909.05858 (2019).
+        """
+        if self.do_apply_repetition_penalty:
+            generated_token_idx = set(past_sequence[0].tolist())
+            for token_idx in generated_token_idx:
+                logits[0, token_idx] /= self.repetition_penalty
+        return logits
+
+    def apply_temperature(self, logits):
+        """ Shape the tokens' distribution through temperature. The higher the value
+        of the temperature, the more skewed towards high probability events the
+        distribution is.
+
+        .. Goodfellow, Ian, Yoshua Bengio, and Aaron Courville. Deep learning.
+        MIT press, 2016.
+        """
+        if self.do_apply_temperature:
+            return logits / self.temperature
+        return logits
+
+    def apply_top_k_filter(self, logits):
+        """ Use the probability distribution of the tokens to determine the set
+        to be sampled from.  Specifically we select the set of size k such that
+        the sum of its items' probabilities is maximum.
+
+        .. Fan, Angela, Mike Lewis, and Yann Dauphin. "Hierarchical neural
+        story generation." arXiv preprint arXiv:1805.04833 (2018).
+        """
+        if self.k > 0:
+            try:
+                indices_to_remove = logits < torch.topk(logits, self.k)[0][..., -1, None]
+                logits[indices_to_remove] = -float("Inf")
+            except RuntimeError:
```

Comment:
Why don't we just take the min between `k` and `logits.size(-1)` instead of throwing an error?

Reply:
Unexpected defaults are a not a good design choice and allow silent bugs. I would advocate for being as defensive as possible on the input values for something that is used in the wild. As a user I'd rather be told I made a mistake (whether it's because of misunderstanding or worse, non intentional), than just having the program carrying on as if everything was fine.

Reply:
I think it's fine in this case.

Reply:
We can do both: send a warning instead of an error & set it to the min of `k` and `logits.size(-1)`